### PR TITLE
implement third react task

### PR DESCRIPTION
**Pull request by AlexChrks**
Deadline: 23.04.2021
Done: 19.04.2021

Screenshot: 
image

Requirements:
  - Each <Card /> component could be deleted from catalogue by clicking on the cross button (❌) which can be placed somewhere on the top of each card. After clicking on the button the array of cards in your <CardsContainer /> should be updated. 

  - <CardsContainer /> component should also additionaly render <CardsCreationForm /> as its children. This form should contain all the required inputs for creating the card (find a way on how to use only 1 handler for all the inputs) and a submit button. By clicking on a submit button the array of cards in your <CardsContainer /> should be updated with the new card data. 

  - All the required handlers for creating/deleting cards should be in the <CardsContainer /> component and passed down to a child components as a props. 

  - Rewrite all classes components to functional components with hooks

### CODE REVIEW

File: react-task-1/src/components/Cards/CardsContainer/CardsContainer.js [hunk 1]
```diff
@@ -1,32 +1,63 @@
...
+  useEffect(() => {
     apiCall().then((data) => {
-      this.setState({ cards: data });
+      setState({ cards: data });
     });
+  }, []);
+
+
+  function deleteEvent(id) {
+    const copyCardsArr = Object.assign([], state.cards);
```
Comment (AliaksandrSemchanka) [inline react-task-1/src/components/Cards/CardsContainer/CardsContainer.js:20]: тээкс, а зачем ты тут делаешь копию?

File: react-task-1/src/components/Cards/CardsContainer/CardsContainer.js [hunk 2]
```diff
@@ -1,32 +1,63 @@
...
+    const filtered = copyCardsArr.filter((el) => !(el.id === id));
```
Comment (AliaksandrSemchanka) [inline react-task-1/src/components/Cards/CardsContainer/CardsContainer.js:21]:
```suggestion
    const filtered = copyCardsArr.filter(el => el.id !== id);
```
так читабельнее 📖

File: react-task-1/src/components/Cards/CardsContainer/CardsContainer.js [hunk 3]
```diff
@@ -1,32 +1,63 @@
...
-      <div className={styles.container}>
-        { cards.map((card) => <Card key={card.id} cardData={card} />)}
-      </div>
+      <div className={styles.noCardsNotification}>No cards yet</div>
     );
   }
+  return (
+    <div className={styles.container}>
+      { cards.map((card) => {
+        return (
```
Comment (AliaksandrSemchanka) [inline react-task-1/src/components/Cards/CardsContainer/CardsContainer.js:52]: давай уберем `return`?

File: react-task-1/src/components/Cards/Card/Card.js [hunk 4]
```diff
@@ -1,9 +1,10 @@
 import React from 'react';
 import styles from './Card.module.scss';
 
-export function Card({ cardData }) {
+export function Card({ cardData, deleteEvent }) {
   return (
     <div className={styles.card}>
+      <button className={styles.btnClose} onClick={deleteEvent}></button>
```
Comment (AliaksandrSemchanka) [inline react-task-1/src/components/Cards/Card/Card.js:7]:
```suggestion
      <button className={styles.btnClose} onClick={deleteEvent} />
```

File: react-task-1/src/components/Cards/Card/Card.js [hunk 5]
```diff
@@ -1,9 +1,10 @@
...
       <img className={styles.images} src={cardData.imageUrl} alt={cardData.type}></img>
```
Comment (AliaksandrSemchanka) [inline react-task-1/src/components/Cards/Card/Card.js:8]:
```suggestion
      <img className={styles.images} src={cardData.imageUrl} alt={cardData.type} />
```

File: react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+      <input placeholder="Title"></input>
+
+      <input placeholder="Type"></input>
+
+      <input placeholder="Image url"></input>
+
+      <input placeholder="Price"></input>
+
+      <input placeholder="Country"></input>
```
Comment (AliaksandrSemchanka) [inline react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js:17]:
```suggestion
      <input placeholder="Country" />
```

File: react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    <form className={styles.creationForm} onSubmit={(e) => { handleSubmit(e) }}>
+      <h1 className={styles.heading}>Add card</h1>
+
+      <input placeholder="Title"></input>
+
+      <input placeholder="Type"></input>
+
+      <input placeholder="Image url"></input>
+
+      <input placeholder="Price"></input>
```
Comment (AliaksandrSemchanka) [inline react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js:15]:
```suggestion
      <input placeholder="Price" />
```

File: react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+export function CardsCreationForm({ handleSubmit }) {
+  return (
+    <form className={styles.creationForm} onSubmit={(e) => { handleSubmit(e) }}>
+      <h1 className={styles.heading}>Add card</h1>
+
+      <input placeholder="Title"></input>
+
+      <input placeholder="Type"></input>
+
+      <input placeholder="Image url"></input>
```
Comment (AliaksandrSemchanka) [inline react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js:13]:
```suggestion
      <input placeholder="Image url" />
```

File: react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+import styles from './CardsCreationForm.module.scss';
+
+export function CardsCreationForm({ handleSubmit }) {
+  return (
+    <form className={styles.creationForm} onSubmit={(e) => { handleSubmit(e) }}>
+      <h1 className={styles.heading}>Add card</h1>
+
+      <input placeholder="Title"></input>
+
+      <input placeholder="Type"></input>
```
Comment (AliaksandrSemchanka) [inline react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js:11]:
```suggestion
      <input placeholder="Type" />
```

File: react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js [hunk 10]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import styles from './CardsCreationForm.module.scss';
+
+export function CardsCreationForm({ handleSubmit }) {
+  return (
+    <form className={styles.creationForm} onSubmit={(e) => { handleSubmit(e) }}>
+      <h1 className={styles.heading}>Add card</h1>
+
+      <input placeholder="Title"></input>
```
[the remark this answers is not in the document]
Comment (AliaksandrSemchanka) [reply react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js:9]: что бы не сохранять карточку с пустыми полями

File: react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js [hunk 11]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import styles from './CardsCreationForm.module.scss';
+
+export function CardsCreationForm({ handleSubmit }) {
+  return (
+    <form className={styles.creationForm} onSubmit={(e) => { handleSubmit(e) }}>
```
Comment (AliaksandrSemchanka) [inline react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js:6]:
```suggestion
    <form className={styles.creationForm} onSubmit={handleSubmit}>
```

File: react-task-1/src/components/Cards/CardsContainer/CardsContainer.js [hunk 12]
```diff
@@ -1,32 +1,63 @@
...
     }
+
+    const stateCopy = Object.assign([], state.cards)
+    stateCopy.push(newObj);
+    setState({cards: stateCopy});
+  }
+
+  const { cards } = state;
+
+  if (!cards.length) {
```
[the remark this answers is not in the document]
Comment (AlexChrks) [reply react-task-1/src/components/Cards/CardsContainer/CardsContainer.js:44]: Improved! Now u can add new card if all cards are deleted

File: react-task-1/src/components/Cards/CardsContainer/CardsContainer.js (hunk 1 above)
Comment (AlexChrks) [reply react-task-1/src/components/Cards/CardsContainer/CardsContainer.js:20]: Right! We can filter state.cards directly without copying.. thanks!

File: react-task-1/src/components/Cards/CardsContainer/CardsContainer.js (hunk 2 above)
Comment (AlexChrks) [reply react-task-1/src/components/Cards/CardsContainer/CardsContainer.js:21]: sure! more readable

File: react-task-1/src/components/Cards/CardsContainer/CardsContainer.js (hunk 3 above)
Comment (AlexChrks) [reply react-task-1/src/components/Cards/CardsContainer/CardsContainer.js:52]: Ok!

Comment (AlexChrks) [reply react-task-1/src/components/Cards/CardsContainer/CardsContainer.js:52]: Done

File: react-task-1/src/components/Cards/Card/Card.js (hunk 4 above)
Comment (AlexChrks) [reply react-task-1/src/components/Cards/Card/Card.js:7]: Done!

File: react-task-1/src/components/Cards/Card/Card.js (hunk 5 above)
Comment (AlexChrks) [reply react-task-1/src/components/Cards/Card/Card.js:8]: Done!

File: react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js (hunk 7 above)
Comment (AlexChrks) [reply react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js:15]: Done!

File: react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js (hunk 8 above)
Comment (AlexChrks) [reply react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js:13]: Done!

File: react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js (hunk 9 above)
Comment (AlexChrks) [reply react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js:11]: Done!

File: react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js (hunk 11 above)
Comment (AlexChrks) [reply react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js:6]: Done!

File: react-task-1/src/components/Cards/CardsContainer/CardsContainer.js [hunk 13]
```diff
@@ -1,32 +1,63 @@
...
-      return (
-        <div className={styles.noCardsNotification}>No cards yet</div>
-      );
+    const newObj = {
+      title: e.target[0].value,
+      type: e.target[1].value,
+      imageUrl: e.target[2].value,
+      price: e.target[3].value,
+      country: e.target[4].value,
+      id: state.cards.length + 1,
```
[the remark this answers is not in the document]
Comment (AlexChrks) [reply react-task-1/src/components/Cards/CardsContainer/CardsContainer.js:34]: Ok, done!

File: react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (AlexChrks) [reply react-task-1/src/components/Cards/CardsCreationForm/CardsCreationForm.js:13]: I left this field required

File: react-task-1/src/components/Cards/CardsContainer/CardsContainer.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (AliaksandrSemchanka) [reply react-task-1/src/components/Cards/CardsContainer/CardsContainer.js:41]: давай еще подумаем, что бы `<CardsCreationForm` в `return()` был один раз

File: react-task-1/src/components/Cards/CardsContainer/CardsContainer.js [hunk 14]
```diff
@@ -1,32 +1,62 @@
...
+    setState({ cards: [...state.cards, cardState] });
+  }
 
-    if (!cards.length) {
-      return (
-        <div className={styles.noCardsNotification}>No cards yet</div>
-      );
-    }
+  function handleChange(event) {
+    console.log(event.target.name);
```
[the remark this answers is not in the document]
Comment (AlexChrks) [reply react-task-1/src/components/Cards/CardsContainer/CardsContainer.js:31]: error handler :)

File: react-task-1/src/components/Cards/CardsContainer/CardsContainer.js [hunk 15]
```diff
@@ -1,32 +1,62 @@
...
 import styles from './CardsContainer.module.scss';
 
-export class CardsContainer extends Component {
+export function CardsContainer() {
 
-  state = {
-    cards: []
-  }
+  const [state, setState] = useState({ cards: [] });
+  const [cardState, setCardState] = useState({});
```
[the remark this answers is not in the document]
Comment (AlexChrks) [reply react-task-1/src/components/Cards/CardsContainer/CardsContainer.js:11]: Sure!

File: react-task-1/src/components/Cards/CardsContainer/CardsContainer.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (AlexChrks) [reply react-task-1/src/components/Cards/CardsContainer/CardsContainer.js:41]: done!

**Pull request merged**